WebSocketClient attempts to use HTTP/2 on servers without SETTINGS_ENABLE_CONNECT_PROTOCOL

**Jetty version(s)**
12.0.23

**Description**

When a Jetty WebSocketClient has HTTP/2 enabled using HttpClientTransportDynamic it attempts to establish HTTP/2 WebSockets on servers that do not support it.

**How to reproduce?**

For example try connecting to `wss://echo.websocket.org` - this server currently supports HTTP/2 for HTML but not for WebSockets:

```java
public class WebSocketHttp2 {

    public static class Listener implements Session.Listener {
        private volatile Session session;

        @Override
        public void onWebSocketOpen(Session session) {
            this.session = session;
            session.demand();
        }
    }

    public static void main(String[] args) throws Exception {
        ClientConnector clientConnector = new ClientConnector();
        ClientConnectionFactory.Info http1 = HttpClientConnectionFactory.HTTP11;
        ClientConnectionFactory.Info http2 = new ClientConnectionFactoryOverHTTP2.HTTP2(new HTTP2Client(clientConnector));
        HttpClient httpClient = new HttpClient(new HttpClientTransportDynamic(clientConnector, http1, http2));
        WebSocketClient client = new WebSocketClient(httpClient);
        client.start();
        client.connect(new Listener(), URI.create("wss://echo.websocket.org")).get();
    }
}
```

Output:

```
Exception in thread "main" java.util.concurrent.ExecutionException: java.io.IOException: protocol_error
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2073)
	at org.example.WebSocketHttp2.main(WebSocketHttp2.java:35)
Caused by: java.io.IOException: protocol_error
	at org.eclipse.jetty.http2.client.transport.internal.HttpReceiverOverHTTP2.lambda$onReset$4(HttpReceiverOverHTTP2.java:242)
	at org.eclipse.jetty.util.thread.Invocable$ReadyTask.run(Invocable.java:177)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.runTask(AdaptiveExecutionStrategy.java:480)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.consumeTask(AdaptiveExecutionStrategy.java:428)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.tryProduce(AdaptiveExecutionStrategy.java:293)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.produce(AdaptiveExecutionStrategy.java:195)
	at org.eclipse.jetty.http2.HTTP2Connection.produce(HTTP2Connection.java:210)
	at org.eclipse.jetty.http2.HTTP2Connection.onFillable(HTTP2Connection.java:157)
```

You can see with `nghttp2` that this server does not advertise support:

```
nghttp -v https://echo.websocket.org/
[  0.087] recv SETTINGS frame <length=24, flags=0x00, stream_id=0>
          (niv=4)
          [SETTINGS_MAX_CONCURRENT_STREAMS(0x03):100]
          [SETTINGS_INITIAL_WINDOW_SIZE(0x04):1048576]
          [SETTINGS_MAX_FRAME_SIZE(0x05):16384]
          [SETTINGS_MAX_HEADER_LIST_SIZE(0x06):262144]
```

Compare to a Jetty server which does:

```
nghttp -v https://localhost:8443
[  0.406] recv SETTINGS frame <length=24, flags=0x00, stream_id=0>
          (niv=4)
          [SETTINGS_MAX_CONCURRENT_STREAMS(0x03):128]
          [SETTINGS_INITIAL_WINDOW_SIZE(0x04):524288]
          [SETTINGS_MAX_HEADER_LIST_SIZE(0x06):131072]
          [SETTINGS_ENABLE_CONNECT_PROTOCOL(0x08):1]
```

**Expected Behavior**

https://datatracker.ietf.org/doc/html/rfc8441#section-3

If the server does not advertise `SETTINGS_ENABLE_CONNECT_PROTOCOL` then the WebSocketClient should dynamically use HTTP/1.1 instead.

```
If a client were to use the provisions of the extended CONNECT method defined in
this document without first receiving a SETTINGS_ENABLE_CONNECT_PROTOCOL parameter, a non-supporting peer
would detect a malformed request and generate a stream error
```

Presumably Firefox is checking for this:

<img width="1125" height="214" alt="Image" src="https://github.com/user-attachments/assets/04e67d4f-969e-425c-b886-c156bd479a57" />

<img width="1188" height="228" alt="Image" src="https://github.com/user-attachments/assets/2af5b601-62f8-44a7-8f35-d9a1efe2c37a" />

Thanks @sbordet I have tried it out, the first issue is that it seems to try IPv6 even though I don't have an IPv6 connection

```
Exception in thread "main" java.util.concurrent.ExecutionException: java.net.SocketException: Could not connect to echo.websocket.org/[2a09:8280:1:0:0:0:37:b5c3]:443
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2073)
	at org.example.WebSocketHttp2.main(WebSocketHttp2.java:35)
Caused by: java.net.SocketException: Could not connect to echo.websocket.org/[2a09:8280:1:0:0:0:37:b5c3]:443
	at org.eclipse.jetty.io.ClientConnector.connect(ClientConnector.java:445)
	at org.eclipse.jetty.io.Transport$Socket.connect(Transport.java:183)
	at org.eclipse.jetty.client.HttpClient.connect(HttpClient.java:629)
	at org.eclipse.jetty.client.AbstractConnectorHttpClientTransport.connect(AbstractConnectorHttpClientTransport.java:65)
	at org.eclipse.jetty.client.HttpClient$1.connect(HttpClient.java:610)
	at org.eclipse.jetty.client.HttpClient$1$1.failed(HttpClient.java:607)
	at org.eclipse.jetty.util.Promise$1.failed(Promise.java:97)
	at org.eclipse.jetty.io.ClientConnector.connectFailed(ClientConnector.java:540)
	at org.eclipse.jetty.io.ClientConnector$ClientSelectorManager.connectionFailed(ClientConnector.java:669)
	at org.eclipse.jetty.io.ManagedSelector$Connect.failed(ManagedSelector.java:972)
	at org.eclipse.jetty.io.ManagedSelector$CreateEndPoint.run(ManagedSelector.java:1066)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:999)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.doRunJob(QueuedThreadPool.java:1229)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1184)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.net.SocketException: Network is unreachable
	at java.base/sun.nio.ch.Net.connect0(Native Method)
	at java.base/sun.nio.ch.Net.connect(Net.java:589)
	at java.base/sun.nio.ch.Net.connect(Net.java:596)
	at java.base/sun.nio.ch.SocketChannelImpl.connect(SocketChannelImpl.java:880)
	at org.eclipse.jetty.io.ClientConnector.connect(ClientConnector.java:423)
	... 14 more
```

Although I'm not sure if this is related to this PR, or some other change in Jetty 12.1.x

I disabled IPv6 using the VM option `-Djava.net.preferIPv4Stack=true` and tried again and got a different error

```java
Exception in thread "main" java.util.concurrent.ExecutionException: java.io.IOException: Cannot find ClientConnectionFactory for protocol null
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2073)
	at org.example.WebSocketHttp2.main(WebSocketHttp2.java:35)
Caused by: java.io.IOException: Cannot find ClientConnectionFactory for protocol null
	at org.eclipse.jetty.client.transport.HttpClientTransportDynamic.lambda$newConnection$4(HttpClientTransportDynamic.java:214)
	at java.base/java.util.Optional.orElseThrow(Optional.java:403)
	at org.eclipse.jetty.client.transport.HttpClientTransportDynamic.newConnection(HttpClientTransportDynamic.java:214)
	at org.eclipse.jetty.io.ssl.SslClientConnectionFactory.newConnection(SslClientConnectionFactory.java:131)
	at org.eclipse.jetty.io.Transport.newConnection(Transport.java:167)
	at org.eclipse.jetty.io.ClientConnector.newConnection(ClientConnector.java:521)
	at org.eclipse.jetty.io.ClientConnector$ClientSelectorManager.newConnection(ClientConnector.java:632)
	at org.eclipse.jetty.io.ManagedSelector.createEndPoint(ManagedSelector.java:388)
	at org.eclipse.jetty.io.ManagedSelector$CreateEndPoint.run(ManagedSelector.java:1058)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:999)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.doRunJob(QueuedThreadPool.java:1229)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1184)
	at java.base/java.lang.Thread.run(Thread.java:1583)
```

<img width="997" height="141" alt="Image" src="https://github.com/user-attachments/assets/4bd6e69e-a6d7-4967-aa1e-2776ab511d1f" />

I'm just running the example code I included in my original issue

--- a/jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/HttpClient.java
+++ b/jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/HttpClient.java
@@ -491,7 +491,7 @@ public class HttpClient extends ContainerLifeCycle implements AutoCloseable
         String scheme = request.getScheme();
         if (!HttpScheme.HTTP.is(scheme) && !HttpScheme.HTTPS.is(scheme) &&
             !HttpScheme.WS.is(scheme) && !HttpScheme.WSS.is(scheme))
-            throw new IllegalArgumentException("Invalid protocol " + scheme);
+            throw new IllegalArgumentException("Invalid URI scheme " + scheme);
         scheme = scheme.toLowerCase(Locale.ENGLISH);
         String host = request.getHost();
         host = host.toLowerCase(Locale.ENGLISH);

--- a/jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/HttpRedirector.java
+++ b/jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/HttpRedirector.java
@@ -155,16 +155,26 @@ public class HttpRedirector
                 if (LOG.isDebugEnabled())
                     LOG.debug("Redirecting to {} (Location: {})", newURI, location);
 
-                if (listener == null)
+                Response.CompleteListener wrapper = result ->
                 {
-                    listener = result ->
+                    if (listener != null)
+                        listener.onComplete(result);
+                    if (result.isFailed())
                     {
-                        if (result.isFailed())
+                        // If the redirect request has already been added to the conversation,
+                        // failures will be notified to the original request by HttpClient, and
+                        // we must not do it, otherwise recursion happens and then StackOverflowError.
+                        // If the redirect request has not been added to the conversation yet,
+                        // we must notify the failure by calling fail(), like we do when we
+                        // cannot redirect (see the else branch below where we cannot redirect
+                        // because the Location header is missing).
+                        Request redirectRequest = result.getRequest();
+                        if (!((HttpRequest)request).getConversation().contains(redirectRequest))
                             fail(result);
-                    };
-                }
+                    }
+                };
 
-                return redirect(request, response, listener, newURI);
+                return redirect(request, response, wrapper, newURI);
             }
             else
             {
@@ -293,10 +303,10 @@ public class HttpRedirector
                 String authority = matcher.group(3);
                 String path = matcher.group(4);
                 String query = matcher.group(5);
-                if (query.length() == 0)
+                if (query.isEmpty())
                     query = null;
                 String fragment = matcher.group(6);
-                if (fragment.length() == 0)
+                if (fragment.isEmpty())
                     fragment = null;
                 try
                 {
@@ -390,6 +400,9 @@ public class HttpRedirector
 
     private void fail(Request request, Throwable requestFailure, Response response, Throwable responseFailure)
     {
+        // This method should only be called to fail the conversation
+        // when the redirect request has not been sent.
+        // See comment in #redirect(Request, Response, CompleteListener).
         HttpConversation conversation = ((HttpRequest)request).getConversation();
         conversation.updateResponseListeners(null);
         conversation.getResponseListeners().emitFailureComplete(new Result(request, requestFailure, response, responseFailure));

--- a/jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/transport/HttpConversation.java
+++ b/jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/transport/HttpConversation.java
@@ -22,6 +22,7 @@ import org.eclipse.jetty.client.Request;
 import org.eclipse.jetty.client.Response;
 import org.eclipse.jetty.util.Attributes;
 import org.eclipse.jetty.util.Promise;
+import org.eclipse.jetty.util.TypeUtil;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -142,6 +143,17 @@ public class HttpConversation extends Attributes.Lazy
     }
 
     /**
+     * <p>Whether the given request is in this conversation.</p>
+     *
+     * @param request the request to test
+     * @return whether the given request is in this conversation
+     */
+    public boolean contains(Request request)
+    {
+        return getExchanges().stream().anyMatch(exchange -> exchange.getRequest() == request);
+    }
+
+    /**
      * <p>Returns the total timeout for the conversation.</p>
      * <p>The conversation total timeout is the total timeout
      * of the first request in the conversation.</p>
@@ -167,6 +179,6 @@ public class HttpConversation extends Attributes.Lazy
     @Override
     public String toString()
     {
-        return String.format("%s[%x]", HttpConversation.class.getSimpleName(), hashCode());
+        return String.format("%s@%x[exchanges=%d]", TypeUtil.toShortName(getClass()), hashCode(), exchanges.size());
     }
 }

--- a/jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/transport/HttpRequest.java
+++ b/jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/transport/HttpRequest.java
@@ -764,15 +764,31 @@ public class HttpRequest implements Request
     @Override
     public void send(Response.CompleteListener listener)
     {
+        Destination destination = resolveDestination(listener);
+        if (destination == null)
+            return;
         try
         {
-            Destination destination = client.resolveDestination(this);
             destination.send(this, listener);
         }
         catch (Throwable x)
         {
+            abort(x);
+        }
+    }
+
+    private Destination resolveDestination(Response.CompleteListener listener)
+    {
+        try
+        {
+            return client.resolveDestination(this);
+        }
+        catch (Throwable x)
+        {
+            // Must notify the listener, since it is not yet associated with this request.
             Result result = new Result(this, x, new HttpResponse(this), null);
             abort(x).thenRun(() -> ResponseListeners.notifyComplete(listener, result));
+            return null;
         }
     }
 

--- a/jetty-core/jetty-client/src/test/java/org/eclipse/jetty/client/HttpClientRedirectTest.java
+++ b/jetty-core/jetty-client/src/test/java/org/eclipse/jetty/client/HttpClientRedirectTest.java
@@ -47,6 +47,7 @@ import static org.hamcrest.MatcherAssert.assertThat;
 import static org.junit.jupiter.api.Assertions.assertArrayEquals;
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertInstanceOf;
 import static org.junit.jupiter.api.Assertions.assertNotNull;
 import static org.junit.jupiter.api.Assertions.assertThrows;
 import static org.junit.jupiter.api.Assertions.assertTrue;
@@ -314,7 +315,7 @@ public class HttpClientRedirectTest extends AbstractHttpClientServerTest
             protected void service(Request request, org.eclipse.jetty.server.Response response) throws Exception
             {
                 response.setStatus(303);
-                response.getHeaders().put("Location", "ssh://localhost:" + connector.getLocalPort() + "/path");
+                response.getHeaders().put("Location", "ssh://localhost:" + connector.getLocalPort() + "/wrong/scheme");
             }
         });
 
@@ -657,6 +658,40 @@ public class HttpClientRedirectTest extends AbstractHttpClientServerTest
                 .timeout(3 * timeout, TimeUnit.MILLISECONDS)
                 .send();
         });
+    }
+
+    @ParameterizedTest
+    @ArgumentsSource(ScenarioProvider.class)
+    public void testAbortRedirectRequest(Scenario scenario) throws Exception
+    {
+        start(scenario, new EmptyServerHandler()
+        {
+            @Override
+            protected void service(Request request, org.eclipse.jetty.server.Response response) throws Throwable
+            {
+                String serverURI = scenario.getScheme() + "://localhost:" + connector.getLocalPort();
+                String target = Request.getPathInContext(request);
+                if ("/initial".equals(target))
+                {
+                    response.setStatus(HttpStatus.SEE_OTHER_303);
+                    response.getHeaders().put(HttpHeader.LOCATION, serverURI + "/redirect");
+                }
+            }
+        });
+
+        client.getRequestListeners().addBeginListener(r ->
+        {
+            if ("/redirect".equals(r.getPath()))
+                r.abort(new HttpRequestException("aborted by test", r));
+        });
+
+        ExecutionException failure = assertThrows(ExecutionException.class, () -> client.newRequest("localhost", connector.getLocalPort())
+            .scheme(scenario.getScheme())
+            .path("/initial")
+            .timeout(5, TimeUnit.SECONDS)
+            .send());
+
+        assertInstanceOf(HttpRequestException.class, failure.getCause());
     }
 
     private void testSameMethodRedirect(final Scenario scenario, final HttpMethod method, int redirectCode) throws Exception

--- a/jetty-core/jetty-client/src/test/java/org/eclipse/jetty/client/HttpClientTest.java
+++ b/jetty-core/jetty-client/src/test/java/org/eclipse/jetty/client/HttpClientTest.java
@@ -94,6 +94,7 @@ import static org.hamcrest.Matchers.notNullValue;
 import static org.junit.jupiter.api.Assertions.assertArrayEquals;
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertInstanceOf;
 import static org.junit.jupiter.api.Assertions.assertNotNull;
 import static org.junit.jupiter.api.Assertions.assertThrows;
 import static org.junit.jupiter.api.Assertions.assertTrue;
@@ -2094,6 +2095,20 @@ public class HttpClientTest extends AbstractHttpClientServerTest
             .headers(h -> h.put("X-Capacity", capacity))
             .timeout(5, TimeUnit.SECONDS)
             .send());
+    }
+
+    @ParameterizedTest
+    @ArgumentsSource(ScenarioProvider.class)
+    public void testRequestWithWrongScheme(Scenario scenario) throws Exception
+    {
+        start(scenario, new EmptyServerHandler());
+
+        ExecutionException failure = assertThrows(ExecutionException.class, () -> client.newRequest("localhost", connector.getLocalPort())
+            .scheme("ssh")
+            .timeout(5, TimeUnit.SECONDS)
+            .send());
+
+        assertInstanceOf(IllegalArgumentException.class, failure.getCause());
     }
 
     private void assertCopyRequest(Request original)

